### PR TITLE
[EVM] Add support of function calls.

### PR DESCRIPTION
**Pull request by PavelKopyl**
- Add support of function calls.
- Add intial Target Transform Info implementation.

### CODE REVIEW

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

```

File: llvm/lib/Target/EVM/EVMISelLowering.cpp [hunk 1]
```diff
@@ -125,6 +125,94 @@ SDValue EVMTargetLowering::LowerFormalArguments(
...
+    // registers.
+    InTys.push_back(In.VT);
+  }
+
+  InTys.push_back(MVT::Other);
+  SDVTList InTyList = DAG.getVTList(InTys);
+  SDValue Res = DAG.getNode(Ins.size() == 0 ? EVMISD::CALL0 : EVMISD::CALL1, DL,
+                            InTyList, Ops);
+
+  for (size_t I = 0; I < Ins.size(); ++I)
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMISelLowering.cpp]: `Ins.size() <= 1`, so do we really need a loop here?

File: llvm/lib/Target/EVM/EVMISelLowering.cpp [hunk 2]
```diff
@@ -125,6 +125,94 @@ SDValue EVMTargetLowering::LowerFormalArguments(
...
+  CLI.IsTailCall = false;
+
+  if (CLI.IsVarArg)
+    fail(DL, DAG, "EVM hasn't implemented variable arguments");
+
+  SmallVectorImpl<ISD::InputArg> &Ins = CLI.Ins;
+  SmallVectorImpl<ISD::OutputArg> &Outs = CLI.Outs;
+  SmallVectorImpl<SDValue> &OutVals = CLI.OutVals;
+
+  for (unsigned I = 0; I < Outs.size(); ++I) {
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMISelLowering.cpp]: Style: ranged for is preferred

**Review by akiramenai — APPROVED** [review]

File: llvm/lib/Target/EVM/EVMInstrInfo.td [hunk 3]
```diff
@@ -94,6 +149,24 @@ let isAsCheapAsAMove = 1 in
 defm COPY_I256
   : I<(outs GPR:$res), (ins GPR:$src), [], "COPY_I256 $res, $src">;
 
+let isCall = 1, hasCtrlDep = 1 in {
+def CALL0
+  : NI<(outs), (ins function_op:$callee, variable_ops),
```
Comment (lialan) [inline llvm/lib/Target/EVM/EVMInstrInfo.td]: `function_op` should be a symbol?

Comment (lialan) [reply llvm/lib/Target/EVM/EVMInstrInfo.td]: or addr

Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMInstrInfo.td]: Actually function_op is an address.
I removed function_op and used here just GPR to make this explicit.  It seems for me that a good model describing function calls in EVM(though we only have jump instructions) is indirect function calls, as after stackification the target address is located om the stack. That's why an address shoud be used there.
This differs from the WebAssemby backed. There a function accepts a symbol as an argument. And also there are disctint insteuctions for indirect call. To be hnest I didn't get what was the reason for that.

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

```

File: llvm/lib/Target/EVM/EVMISelLowering.cpp (hunk 1 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMISelLowering.cpp]: Thank you. Fixed.

File: llvm/lib/Target/EVM/EVMISelLowering.cpp (hunk 2 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMISelLowering.cpp]: Fixed.

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

```

**Pull request merged**